### PR TITLE
Feature/axes renderable conversion

### PR DESCRIPTION
**Pull request by hydra41**
Converted Axes renderable from the triangle to the Axes.
    - Updated the PassThru shaders to version 450 and modified them to compile
    - Fixed an issue with the DescriptorPool not being setup properly (incorrect MaxSets value was causing a PPOL_OUT_OF_MEMORY error)
    - Fixed the rotation matrix from the VisualProcesser to default to the Identity matrix until the camera is setup correctly

### CODE REVIEW

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java [hunk 1]
```diff
@@ -649,26 +658,35 @@ public int InitVKDescriptorPool(MemoryStack stack, CVKSynchronizedDescriptorType
                 if (count > 0) {
                     VkDescriptorPoolSize poolSize = pPoolSizes.get(iPoolSize++);
                     poolSize.type(iType);
-                    int size = CalculateDescriptorPoolSizeForType(iType, 
-                                                                  poolDescriptorTypeCounts[iType],
-                                                                  count); 
+                    
+                    // Leaner and meaner version. Just allocate what we need. This could get out of control with 1000s of nodes.
+                    int size = count;
+                    //int size = CalculateDescriptorPoolSizeForType(iType,
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java]: Why did you neuter that code path?  Sizing the pool to the current size means we'll be recreating the pool every time a new icon is added.  That seems wasteful.  Also I'm not sure what happens to all the existing descriptors when the pool is recreated.  Can you test that (later if fine) on the pool recreation please?

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/shaders/PassThru.vs [hunk 2]
```diff
@@ -1,17 +1,19 @@
-#version 330 core
+#version 450
 
 
 // === UNIFORMS
-layout(std140) uniform UniformBlock {
-    mat4 mvpMatrix;
-};
+layout(std140, binding = 0) uniform UniformBlock {
+    mat4 mvMatrix;
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/shaders/PassThru.vs]: Why did you rename this?  Given there is no project to screen space happening in the geometry shader I think this might well be a model-view-projection matrix.

**Review by TheTimurid — CHANGES_REQUESTED** [review]

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/shaders/PassThru.vs (hunk 2 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/shaders/PassThru.vs]: renamed back to mvp

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java (hunk 1 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java]: restated the old code path

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java [hunk 3]
```diff
@@ -640,35 +668,49 @@ public int InitVKDescriptorPool(MemoryStack stack, CVKSynchronizedDescriptorType
         
         // Do we have anything to render?
         int allTypesCount = desiredPoolDescriptorTypeCounts.NumberOfDescriptorTypes();
+        int maxSets = 0;
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java:671]: Is this now unused?

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 4]
```diff
@@ -91,72 +89,98 @@
...
+    private static final float LEN = 0.5f;
+    private static final float HEAD = 0.05f;
+    private static final int AXES_OFFSET = 50;
+    private static final Vector4f XCOLOR = new Vector4f(1, 0.5f, 0.5f, 0.75f);
+    private static final Vector4f YCOLOR = new Vector4f(0.5f, 1, 0.5f, 0.75f);
+    private static final Vector4f ZCOLOR = new Vector4f(0, 0.5f, 1, 0.75f);
+    private static final Vector3f ZERO_3F = new Vector3f(0, 0, 0);
+    private static final Matrix44f IDENTITY_44F = Matrix44f.identity();
+
+    private static final int NUMBER_OF_VERTICES = 3 * 2 + 4 + 4 + 4 + 4 + 6 + 6;
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:139]: I know this isn't your code, but I think it could use a comment explaining why there are so many verts for this lineline.

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 5]
```diff
@@ -242,41 +266,40 @@ private void DestroyCommandBuffers() {
     
     
     private void DestroyPipeline() {
-        if (null != vertUniformBuffers && vertUniformBuffers.size() > 0) {
-            for (int i = 0; i < vertUniformBuffers.size(); ++i) {
-                vkDestroyBuffer(cvkDevice.GetDevice(), vertUniformBuffers.get(i).GetBufferHandle(), null);
-                vkFreeMemory(cvkDevice.GetDevice(), vertUniformBuffers.get(i).GetMemoryBufferHandle(), null);
+        if (null != vertexUniformBuffers && vertexUniformBuffers.size() > 0) {
+            for (int i = 0; i < vertexUniformBuffers.size(); ++i) {
+                vkDestroyBuffer(cvkDevice.GetDevice(), vertexUniformBuffers.get(i).GetBufferHandle(), null);
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:271]: The method is called destroy pipeline but it is also destroying the vertex UB.  The creation and destruction of UB should be in separate methods.

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 6]
```diff
@@ -285,6 +308,236 @@ public int Init(CVKDevice cvkDevice) {
...
+        return 25.0f / yScale;
+    } 
+    
+    private int CreateUniformBuffers(MemoryStack stack, CVKSwapChain cvkSwapChain) {
+        int ret = VK_SUCCESS;
+     
+        // TODO_TT: investigate a frames in flight < imageCount approach
+        int imageCount = cvkSwapChain.GetImageCount();        
+        
+        // LIFTED FROM AxeaRenerable.reshape(...)
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:329]: typo, axea

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 7]
```diff
@@ -285,6 +308,236 @@ public int Init(CVKDevice cvkDevice) {
...
+        Graphics3DUtilities.project(unitPosition, IDENTITY_44F, viewport, proj2);
+        final float yScale = proj2.a[1] - proj1.a[1];
+
+        return 25.0f / yScale;
+    } 
+    
+    private int CreateUniformBuffers(MemoryStack stack, CVKSwapChain cvkSwapChain) {
+        int ret = VK_SUCCESS;
+     
+        // TODO_TT: investigate a frames in flight < imageCount approach
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:326]: Can probably remove this as we have opted for frames = images.

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 8]
```diff
@@ -285,6 +308,236 @@ public int Init(CVKDevice cvkDevice) {
...
+        // so they should probably be allocated as VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT and staged
+        // to once to fill them (staging buffer this is host visible then copied to the device local)
+        //for (int j = 0; j < imageCount; ++j) {   
+        CVKBuffer vertexBuffer = CVKBuffer.Create(cvkDevice, 
+                                                        size,
+                                                        VK_BUFFER_USAGE_VERTEX_BUFFER_BIT,
+                                                        VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
+        vertexBuffers.add(vertexBuffer);
+
+        PointerBuffer data = stack.mallocPointer(1);
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:529]: pData

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 9]
```diff
@@ -285,6 +308,236 @@ public int Init(CVKDevice cvkDevice) {
...
+        // axesBatch.stage(colorTarget, ZCOLOR);
+        // axesBatch.stage(vertexTarget, HEAD, -HEAD, LEN + HEAD);
+        vertices[i++] = new Vertex(new Vector3f(HEAD, -HEAD, LEN + HEAD), ZCOLOR);
+        
+         
+        //TODO_TT: most if not all of Constellation's vertex buffers won't change after creation
+        // so they should probably be allocated as VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT and staged
+        // to once to fill them (staging buffer this is host visible then copied to the device local)
+        //for (int j = 0; j < imageCount; ++j) {   
+        CVKBuffer vertexBuffer = CVKBuffer.Create(cvkDevice,
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:523]: cvkVertexBuffer?  I feel like cvk is good for denoting member variables of types we've created, but not sure it is needed for locals.  Probably should though for consistency, can change later.

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 10]
```diff
@@ -313,241 +575,165 @@ public int CreatePipeline(CVKSwapChain cvkSwapChain) {
...
-            fragShaderStageInfo.module(fragShaderModule);
-            fragShaderStageInfo.pName(entryPoint);
-
-            // ===> VERTEX STAGE <===
-
-            VkPipelineVertexInputStateCreateInfo vertexInputInfo = VkPipelineVertexInputStateCreateInfo.callocStack(stack);
-            vertexInputInfo.sType(VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO);
-            vertexInputInfo.pVertexBindingDescriptions(Vertex.getBindingDescription());         // From Vertex struct
-            vertexInputInfo.pVertexAttributeDescriptions(Vertex.getAttributeDescriptions());    // From Vertex struct
+            ret = CreateUniformBuffers(stack, cvkSwapChain);
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:578]: Call from whatever is calling createPipeline

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 11]
```diff
@@ -313,241 +575,165 @@ public int CreatePipeline(CVKSwapChain cvkSwapChain) {
...
-            rasterizer.frontFace(VK_FRONT_FACE_CLOCKWISE);
-            rasterizer.depthBiasEnable(false);
-
-            // ===> MULTISAMPLING <===
-
-            VkPipelineMultisampleStateCreateInfo multisampling = VkPipelineMultisampleStateCreateInfo.callocStack(stack);
-            multisampling.sType(VK_STRUCTURE_TYPE_PIPELINE_MULTISAMPLE_STATE_CREATE_INFO);
-            multisampling.sampleShadingEnable(false);
-            multisampling.rasterizationSamples(VK_SAMPLE_COUNT_1_BIT);
+            ret = CreateDescriptorSets(stack, cvkSwapChain);
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:581]: Call elsewhere.  Also...in recreateSwapchain put an image change check to see if you even need to recreate everything.  See FPS (not sure I have pushed it, will check).

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 12]
```diff
@@ -313,241 +575,165 @@ public int CreatePipeline(CVKSwapChain cvkSwapChain) {
...
+                rasterizer.depthClampEnable(false);
+                rasterizer.rasterizerDiscardEnable(false);
+                rasterizer.polygonMode(VK_POLYGON_MODE_FILL);
+                rasterizer.lineWidth(1.0f);
+                rasterizer.cullMode(VK_CULL_MODE_BACK_BIT);
+                rasterizer.frontFace(VK_FRONT_FACE_COUNTER_CLOCKWISE);
+                rasterizer.depthBiasEnable(false);
+
+                // ===> MULTISAMPLING <===
+                VkPipelineMultisampleStateCreateInfo multisampling = VkPipelineMultisampleStateCreateInfo.callocStack(stack);
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:658]: This should be vkMultiSampling etc but like I mentioned above, not sure how much value these add for local variables.

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 13]
```diff
@@ -622,10 +800,12 @@ public static int LoadShaders(CVKDevice cvkDevice) {
         int ret = VK_SUCCESS;
 
         try{
-            vertShaderSPIRV = compileShaderFile(CVKShaderPlaceHolder.class, "17_shader_vertexbuffer.vert", VERTEX_SHADER);
-            fragShaderSPIRV = compileShaderFile(CVKShaderPlaceHolder.class, "17_shader_vertexbuffer.frag", FRAGMENT_SHADER);
-
+            vertShaderSPIRV = compileShaderFile(CVKShaderPlaceHolder.class, "PassThru.vs", VERTEX_SHADER);
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:803]: Wait....we're compiling shaders from disk every frame????  It's good to have that code path but it should only get called if the .spv is missing or out of date.

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java [hunk 14]
```diff
@@ -643,58 +823,99 @@ public int DisplayUpdate(CVKSwapChain cvkSwapChain, int frameIndex) {
...
+    }         
+    
+    
+    private int CreateDescriptorSets(MemoryStack stack, CVKSwapChain cvkSwapChain) {
+        int ret;
+     
+        int imageCount = cvkSwapChain.GetImageCount();
+
+        // Create descriptor sets
+        LongBuffer layouts = stack.mallocLong(imageCount);
```
Comment (TheTimurid) [inline CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:871]: pLayouts

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java (hunk 3 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/CVKSwapChain.java:671]: yep deleted

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java (hunk 5 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:271]: Updated with separate functions

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java (hunk 6 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:329]: Fixed

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java (hunk 7 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:326]: Removed

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java (hunk 4 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:139]: Added comment explaining the verts

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java (hunk 13 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:803]: It's called once at the beginning of the graph creation

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java (hunk 14 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:871]: changed

File: CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java (hunk 12 above)
Comment (hydra41) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:658]: not much considering they are used in only a few lines right next to each other, I'm going to leave this

Comment (TheTimurid) [reply CoreVulkanDisplay/src/au/gov/asd/tac/constellation/visual/vulkan/renderables/CVKAxesRenderable.java:658]: Yep fair enough
